### PR TITLE
Handle when the energy counter (energy_uj) wraps around (reaches max_…

### PR DESCRIPTION
**Pull request by gcorrall**
Handles when the energy counter (energy_uj) wraps around (reaches max_energy_range_uj) when using Intel RAPL.

### CODE REVIEW

File: src/intel_rapl_sysfs.adb [hunk 1]
```diff
@@ -87,4 +87,26 @@ package body Intel_RAPL_sysfs is
             --Put_Line ("Error reading file " & Package_Name & " for Intel RAPL.");
     end;
 
+    procedure Get_Max_Energy_Range (RAPL_Data : in out Intel_RAPL_Data; Package_Name : in String) is
+        F_Name : File_Type; -- File handle
+        Folder_Name : constant String := "/sys/class/powercap/intel-rapl/"; -- Folder prefix for file to read
+    begin
+        if (Package_Name = "psys") then
+            Open (F_Name, In_File, Folder_Name & "intel-rapl:1/max_energy_range_uj");
+            RAPL_Data.max_energy_range := Long_Float'Value (Get_Line (F_Name)) / 1000000.0;
```
Comment (adelnoureddine) [inline src/intel_rapl_sysfs.adb:96]: I'd rather have distinct variables for psys and pkg

File: src/intel_rapl_sysfs.ads [hunk 2]
```diff
@@ -26,6 +26,10 @@ package Intel_RAPL_sysfs is
             psys_supported : Boolean := False; -- if system supports psys
             pkg_supported : Boolean := False; -- if system supports pkg 0
             dram_supported : Boolean := False; -- if system support dram 0
+
+            -- Data to store max energy ranges
+            max_energy_range : Long_Float := 0.0; -- max_energy_range_uj
```
Comment (adelnoureddine) [inline src/intel_rapl_sysfs.ads:31]: I'd rather have distinct variables for psys and pkg, so pkg_max_energy_range and psys_max_energy_range

File: src/powerjoular.adb [hunk 3]
```diff
@@ -257,6 +267,19 @@ begin
...
             RAPL_Energy := RAPL_After.total_energy - RAPL_Before.total_energy;
+
+            if RAPL_Before.Dram_supported then
+                if RAPL_Before.dram > RAPL_After.dram then
+                    -- dram has wrapped
+                    RAPL_Energy := RAPL_Energy + RAPL_Before.dram_max_energy_range;
+                end if;
+            end if;
+
+            if RAPL_Before.total_energy > RAPL_After.total_energy then
```
Comment (adelnoureddine) [inline src/powerjoular.adb:278]: The way we're calculating power with RAPL is:
- if psys, then provide psys value only (hence here to add max range for psys only)
- if pkg, then add pkg + dram (if dram available)

**Comment by gcorrall** [comment]
Thanks for the review - I've now attempted to address the points you raise.

**Review by adelnoureddine — APPROVED** [review]

**Comment by adelnoureddine** [comment]
Thanks @gcorrall , I'll have a few quick tests on my machine, and then merge it. All good for me 😃

**Pull request merged**